Source/WebCore/page/Page.cpp

```diff
@@ -2124,11 +2124,16 @@ void Page::setDebugger(JSC::Debugger* debugger)
         frame->windowProxy().attachDebugger(m_debugger);
 }
 
-StorageNamespace* Page::sessionStorage(bool optionalCreate)
+StorageNamespace* Page::sessionStorage(SecurityOrigin& topOrigin, bool optionalCreate)
 {
-    if (!m_sessionStorage && optionalCreate) {
-        ASSERT(m_settings->sessionStorageQuota() != StorageMap::noQuota);
-        m_sessionStorage = m_storageNamespaceProvider->createSessionStorageNamespace(*this, m_settings->sessionStorageQuota());
+    if (!m_sessionStorage) {
+        if (optionalCreate) {
+            ASSERT(m_settings->sessionStorageQuota() != StorageMap::noQuota);
+            m_sessionStorage = m_storageNamespaceProvider->createSessionStorageNamespace(topOrigin, *this, m_settings->sessionStorageQuota());
+        }
+    } else {
+        if (m_sessionStorage->topLevelOrigin() && !m_sessionStorage->topLevelOrigin()->isSameOriginDomain(topOrigin))
+            m_sessionStorage = m_storageNamespaceProvider->createSessionStorageNamespace(topOrigin, *this, m_settings->sessionStorageQuota());
```

Comment:
Should we check optionalCreate here too?

Reply:
Yes, and this wound up simplifying the logic, too. `optionalCreate` is now passed all the way down to `StorageNamespaceImpl` where a new namespace is created - now, it's only created if `optionalCreate` is true.